e2e/ls_test.go

```diff
@@ -391,3 +393,204 @@ func TestListS3ObjectsWithDashH(t *testing.T) {
+
+	createBucket(t, s3client, bucket)
+
+	for _, filename := range filenames {
+		putFile(t, s3client, bucket, filename, content)
+
+	}
+
+	cmd := s5cmd("ls", "--exclude", excludePattern, "s3://"+bucket)
+	result := icmd.RunCmd(cmd)
+
+	result.Assert(t, icmd.Success)
+
+	assertLines(t, result.Stdout(), map[int]compareFunc{
+		0: match(`DIR a/`),
+		1: match(`file.py`),
+	}, trimMatch(dateRe), alignment(true))
+}
+
+// ls --exclude ".txt" --exclude ".py" s3://bucket
+func TestListS3ObjectsWithExcludeFilters(t *testing.T) {
+	t.Parallel()
+
+	bucket := s3BucketFromTestName(t)
+
+	const (
+		content         = "content"
+		excludePattern1 = "*.txt"
+		excludePattern2 = "*.py"
+	)
+
+	filenames := []string{
+		"file.txt",
+		"file.py",
+		"hasan.txt",
+		"a/try.txt",
+		"a/try.py",
+		"a/file.c",
+		"file2.txt",
+	}
+
+	s3client, s5cmd, cleanup := setup(t)
+	defer cleanup()
+
+	createBucket(t, s3client, bucket)
+
+	for _, filename := range filenames {
+		putFile(t, s3client, bucket, filename, content)
+
+	}
+
+	cmd := s5cmd("ls", "--exclude", excludePattern1, "--exclude", excludePattern2, "s3://"+bucket)
+	result := icmd.RunCmd(cmd)
+
+	result.Assert(t, icmd.Success)
+
+	assertLines(t, result.Stdout(), map[int]compareFunc{
+		0: match(`DIR a/`),
+	}, trimMatch(dateRe), alignment(true))
+}
+
+// ls --exclude "" s3://bucket
+func TestListS3ObjectsWithExcludeFilterEmpty(t *testing.T) {
+	t.Parallel()
+
+	bucket := s3BucketFromTestName(t)
+
+	const content = "content"
+	const excludePattern = ""
+	filenames := []string{
+		"file.txt",
+		"file.py",
+		"a/try.txt",
+		"a/try.py",
+		"a/file.c",
+		"file2.txt",
+	}
+
+	s3client, s5cmd, cleanup := setup(t)
+	defer cleanup()
+
+	createBucket(t, s3client, bucket)
+
+	for _, filename := range filenames {
+		putFile(t, s3client, bucket, filename, content)
+
+	}
+
+	cmd := s5cmd("ls", "--exclude", excludePattern, "s3://"+bucket)
+	result := icmd.RunCmd(cmd)
+
+	result.Assert(t, icmd.Success)
+
+	assertLines(t, result.Stdout(), map[int]compareFunc{
+		0: match(`DIR a/`),
+		1: match(`file.py`),
+		2: match("file.txt"),
+		3: match("file2.txt"),
+	}, trimMatch(dateRe), alignment(true))
+}
+
+// ls --exclude "main*" directory/
+func TestListLocalFilesWithExcludeFilter(t *testing.T) {
+	t.Parallel()
+
+	_, s5cmd, cleanup := setup(t)
+	defer cleanup()
+
+	const excludePattern = "main*"
+	folderLayout := []fs.PathOp{
+		fs.WithDir(
+			"main",
+			fs.WithFile("try.txt", "this is a txt file"),
+		),
+		fs.WithFile("file1.txt", "this is the first test file"),
+		fs.WithFile("main.py", "this is a python file"),
+		fs.WithFile("main.c", "this is a c file"),
+		fs.WithFile("main.txt", "this is a txt file"),
+		fs.WithFile("main2.txt", "this is a txt file"),
+		fs.WithFile("readme.md", "this is a readme file"),
+	}
+
+	workdir := fs.NewDir(t, t.Name(), folderLayout...)
+	defer workdir.Remove()
+	srcpath := workdir.Path()
+	srcpath = filepath.ToSlash(srcpath)
+
+	cmd := s5cmd("ls", "--exclude", excludePattern, srcpath)
```

Comment:
`srcpath` does not end with `/`.

Reply:
As with my copy comments, there are there cases I can think of with local directories;
* dir (which is why storage.Stat call was needed.)
* dir/
* dir/*

Reply:
added test cases in other  [PR](https://github.com/peak/s5cmd/pull/334).